src/dataflow/src/source/postgres/metrics.rs

```diff
@@ -7,51 +7,106 @@
 impl PgSourceMetrics {
-    pub(super) fn new(base_metrics: &SourceBaseMetrics, source_id: SourceInstanceId) -> Self {
-        let labels = &[source_id.to_string()];
+    pub(super) fn new(
+        base_metrics: &SourceBaseMetrics,
+        source_id: SourceInstanceId,
+        pg_source: &PostgresSourceDetails,
+    ) -> Self {
         let pg_metrics = &base_metrics.postgres_source_specific;
+        let source_label = &[source_id.to_string()];
+        let ops = vec![PgOp::INSERT, PgOp::UPDATE, PgOp::DELETE];
+        let table_metrics = HashMap::from_iter(pg_source.tables.iter().map(|table| {
+            (
+                table.relation_id,
+                HashMap::from_iter(ops.iter().map(|o| {
+                    let labels = &[
+                        source_id.to_string(),
+                        table.namespace.clone(),
+                        table.name.clone(),
+                        o.into(),
+                    ];
+                    (
+                        *o,
+                        pg_metrics
+                            .dml_messages
+                            .get_delete_on_drop_counter(labels.to_vec()),
+                    )
+                })),
+            )
+        }));
+
         Self {
-            inserts: pg_metrics
-                .insert_messages
-                .get_delete_on_drop_counter(labels.to_vec()),
-            updates: pg_metrics
-                .update_messages
-                .get_delete_on_drop_counter(labels.to_vec()),
-            deletes: pg_metrics
-                .delete_messages
-                .get_delete_on_drop_counter(labels.to_vec()),
+            table_ops: table_metrics,
+            tables: pg_metrics
+                .tables_in_publication
+                .get_delete_on_drop_gauge(source_label.to_vec()),
+            lsn: pg_metrics
+                .wal_lsn
+                .get_delete_on_drop_gauge(source_label.to_vec()),
             ignored: pg_metrics
                 .ignored_messages
-                .get_delete_on_drop_counter(labels.to_vec()),
-            total: pg_metrics
-                .total_messages
-                .get_delete_on_drop_counter(labels.to_vec()),
+                .get_delete_on_drop_counter(source_label.to_vec()),
             transactions: pg_metrics
                 .transactions
-                .get_delete_on_drop_counter(labels.to_vec()),
-            tables: pg_metrics
-                .tables_in_publication
-                .get_delete_on_drop_gauge(labels.to_vec()),
-            lsn: pg_metrics.wal_lsn.get_delete_on_drop_gauge(labels.to_vec()),
+                .get_delete_on_drop_counter(source_label.to_vec()),
         }
     }
+
+    pub fn op(&self, table_id: u32, op: PgOp) {
+        self.table_ops
+            .get(&table_id)
+            .and_then(|t| t.get(&op))
+            .and_then(|ctr| Some(ctr.inc()));
+    }
+
+    pub fn lsn(&self, lsn: u64) {
+        self.lsn.set(lsn);
+    }
+
+    pub fn tables(&self) {
```

Comment:
I find this naming very confusing, can it be `tables_inc`, and similarly for the others?